source/lib.ts

```diff
@@ -0,0 +1,326 @@
+// /source/lib.ts
+// The option parser and rate limiting middleware
+
+import Express from 'express'
+
+import {
+	Options,
+	AugmentedRequest,
+	RateLimitRequestHandler,
+	LegacyStore,
+	Store,
+	IncrementResponse,
+} from './types.js'
+import MemoryStore from './memory-store.js'
+
+/**
+ * Type guard to check if a store is legacy store.
+ *
+ * @param store {LegacyStore | Store} - The store to check
+ *
+ * @return {boolean} - Whether the store is a legacy store
+ */
+const isLegacyStore = (store: LegacyStore | Store): store is LegacyStore =>
+	typeof (store as LegacyStore).incr === 'function'
```

Comment:
perhaps we should only consider it as legacy if it has `incr` and it *doesn't* have `increment` - that way store authors can keep `incr` around for backwards-compatibility, while also using the new API. 

Or, perhaps we just add a `apiVersion` field to the `Store` interface - if it's unset or <2, then do the upgrade. If it's >= 2, don't. I don't know.. maybe I'm overthinking this one...

Reply:
> Perhaps we should only consider it as legacy if it has incr and it doesn't have increment - that way store authors can keep incr around for backwards-compatibility, while also using the new API.

Good point, I'll add this.

> Or, perhaps we just add a apiVersion field to the Store interface - if it's unset or <2, then do the upgrade. If it's >= 2, don't. I don't know.. maybe I'm overthinking this one...

A version number sounds a bit overkill, but it might be a good idea if there are a lot of changes to the `Store` interface in the future...